analyzers/src/SonarAnalyzer.CSharp/Rules/ImplementIDisposableCorrectly.cs

```diff
@@ -37,110 +37,119 @@ namespace SonarAnalyzer.Rules.CSharp
     [Rule(DiagnosticId)]
     public sealed class ImplementIDisposableCorrectly : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S3881";
+        private const string DiagnosticId = "S3881";
         private const string MessageFormat = "Fix this implementation of 'IDisposable' to conform to the dispose pattern.";
 
-        private static readonly ISet<SyntaxKind> notAllowedDisposeModifiers = new HashSet<SyntaxKind>
+        private static readonly ISet<SyntaxKind> NotAllowedDisposeModifiers = new HashSet<SyntaxKind>
         {
             SyntaxKind.VirtualKeyword,
             SyntaxKind.AbstractKeyword
         };
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
-        protected override void Initialize(SonarAnalysisContext context)
-        {
+        protected override void Initialize(SonarAnalysisContext context) =>
             context.RegisterSyntaxNodeActionInNonGenerated(c =>
-            {
-                var classDeclaration = (ClassDeclarationSyntax)c.Node;
-
-                var checker = new DisposableChecker(classDeclaration, c.SemanticModel);
-                var locations = checker.GetIssueLocations();
-                if (locations.Any())
                 {
-                    c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
-                        locations.ToAdditionalLocations(),
-                        locations.ToProperties()));
-                }
-            },
-            SyntaxKind.ClassDeclaration);
-        }
+                    if (c.ContainingSymbol.Kind != SymbolKind.NamedType)
+                    {
+                        return;
+                    }
+
+                    var typeDeclarationSyntax = (TypeDeclarationSyntax)c.Node;
+                    var declarationIdentifier = typeDeclarationSyntax.Identifier;
+                    var checker = new DisposableChecker(typeDeclarationSyntax.BaseList,
+                        declarationIdentifier,
```

Comment:
pedantic: the parameters should be aligned one under the other. Rider has a setting for this which is very helpful.